wordpress-mobile/gutenberg-mobile
# Fix scrolling anomalies when splitting a paragraph or appending text to a long paragraph

To Fix: https://github.com/wordpress-mobile/gutenberg-mobile/issues/583

**Cases solved:**

1. appending text to a long paragraph:

Before:

![scrolling-new-line](https://user-images.githubusercontent.com/9772967/51272790-63818980-19cb-11e9-994e-864299b7ea1f.gif)

After:

![glitch-followcaret1](https://user-images.githubusercontent.com/5032900/52593473-1a054c80-2e5a-11e9-98e4-70adf83b8ee6.gif)


2. splitting a long paragraph, scrolling down and never coming back issue

Before:

![glitch1](https://user-images.githubusercontent.com/5032900/52592925-b3336380-2e58-11e9-8173-a11ceb37f829.gif)

After:

![glitch2](https://user-images.githubusercontent.com/5032900/52593146-4cfb1080-2e59-11e9-8944-f1068b74a713.gif)


**Cases couldn't be solved:**

Merge/split glitch issues couldn't be solved because they are caused by RN layer, whenever content size changes content offset is also **sometimes** adjusted accordingly (only under some conditions, see RCTScrollView's calculateOffsetForContentSize:(CGSize)newContentSize to see exactly what conditions, long paragraphs don't have it always) If we could just delete the call for calculateOffsetForContentSize then everything could have worked perfectly(I tried and it worked perfectly). But, currently there's no way to do it, so I am going to open a ticket to RN repo about it, it'd be good to be able to turn it off. They put a comment explaining why they are calling this method:

>    // When contentSize is set manually, ScrollView internals will reset
>    // contentOffset to  {0, 0}. Since we potentially set contentSize whenever
>    // anything in the ScrollView updates, we workaround this issue by manually
>    // adjusting contentOffset whenever this happens

maybe this was needed in old iOS versions, but with iOS 12 it looks like we don't need that workaround imho.
BTW I tried to make lots of workarounds for it but had no chance. Any idea is welcomed.

I should underline this, the problem with merging and splitting blocks is not related with the length of the block's content, sometimes long blocks split/merge pretty good, other times short blocks can have glitch. It is more related with the logic inside calculateOffsetForContentSize.

WPiOS PR: https://github.com/wordpress-mobile/WordPress-iOS/pull/10968

Gutenberg PR: https://github.com/WordPress/gutenberg/pull/13705
react-native-keyboard-aware-scroll-view PR: https://github.com/wordpress-mobile/react-native-keyboard-aware-scroll-view/pull/4


**To Test**

Prerequisites
 
**Example app**

- Run `yarn install` <<important
- Run `yarn start`
- `yarn ios`
- `yarn android`

**WPiOS**

Checkout the PRs branch to any arbitrary folder and cd .. to it
run yarn install, yarn start
Checkout the WPiOS branch in the [WPiOS PR](https://github.com/wordpress-mobile/WordPress-iOS/pull/10968)
Run `rake dependencies`
Open XCode
Clean build folder on Xcode, and then run the app
 
**WPAndroid**

open grade.properties at WordPress-Android folder
add wp.BUILD_GUTENBERG_FROM_SOURCE = true to grade.properties
checkout the PRs branch in the subrepo of WordPress-Android repo
cd to WordPress-Android/libs/gutenberg-mobile
run yarn install, yarn start
yarn wpandroid on a separate terminal in the same directory

Test 1 - Append new text to a very long block
- Prepare a paragraph block that contains text larger than the viewable area
- Append new text at the end and until caret position goes under the keyboard
- Verify that the scrolling glitch problem doesn't occur as described in the [comment](https://github.com/wordpress-mobile/gutenberg-mobile/issues/463#issuecomment-455103187) here.

Test 2 - Insert new text to a very long block
- Prepare a paragraph block that contains text larger than the viewable area
- Choose a position in the middle and insert new text until caret position goes under the keyboard
- Verify that the scrolling glitch problem doesn't occur 

Test 3 - Splitting any long paragraph
- Try splitting any long paragraph in the first line
- Verify that caret won't go out of the viewable area

Test 4 - Test if we didn't affect existing functionality

( Repeat with both heading & paragraph blocks )

- Check if we don't have trouble in adjusting the block height as some text is added or removed
- Check that we have no problem focusing to the recently created block
- Check basic functionality on toolbar like bold/italic etc
- Check basic functionality on block toolbar like up/bottom arrows etc
